core/src/test/java/org/apache/calcite/sql/type/RelDataTypeSystemTest.java

```diff
@@ -153,6 +153,15 @@ private static final class CustomTypeSystem extends RelDataTypeSystemImpl {
     assertEquals(SqlTypeName.DOUBLE, dataType.getSqlTypeName());
   }
 
+  @Test void testLeastRestrictiveUsesMeasureElementType() {
+    RelDataType innerType = TYPE_FACTORY.createSqlType(SqlTypeName.DOUBLE);
+    RelDataType operand1 = TYPE_FACTORY.createMeasureType(innerType);
+    RelDataType operand2 = TYPE_FACTORY.createSqlType(SqlTypeName.INTEGER);
+    RelDataType dataType = SqlLibraryOperators.IFNULL
+        .inferReturnType(TYPE_FACTORY, Lists.newArrayList(operand1, operand2));
+    assertEquals(dataType, operand1);
```

Comment:
`assertEquals(dataType, operand1)` ensures we have a `MEASURE<DOUBLE>`. Is that desired? I think yes, but want to consider if the return type should be the `innerType` of the measure?

Reply:
I think it makes sense for the return type to be a `MEASURE`, for instance if you had `IFNULL(AVG(x), 0)` and `AVG(x)` is not null, the return type should be a `MEASURE<DOUBLE>`. What do you think?

Reply:
Please use `assertThat` rather than `assertEquals`.

I think the measure wrapper should be removed. The type should be `DOUBLE`.

Add a comment explaining what the test does, e.g. "IFNULL(MEASURE<DOUBLE>, INTEGER) should have type "DOUBLE". Much easier for future maintainers to review. Also add an `<a>` link.

Reply:
Done